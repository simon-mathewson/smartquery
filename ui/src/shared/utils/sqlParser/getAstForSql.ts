import type { Connection } from '@/connections/types';
import { castArray } from 'lodash';
import { getParserOptions, getSqlParser } from './utils';

export const getAstForSql = async (props: {
  engine: Connection['engine'];
  skipLogging?: boolean;
  statement: string;
}) => {
  const { engine, statement, skipLogging } = props;

  try {
    const sqlParser = await getSqlParser();
    const ast = sqlParser.astify(statement, getParserOptions(engine));
    return castArray(ast)[0];
  } catch (error) {
    if (!skipLogging) {
      console.error(error);
    }
    return null;
  }
};
