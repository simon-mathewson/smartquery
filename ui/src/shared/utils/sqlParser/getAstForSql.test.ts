import { describe, expect, test } from 'vitest';
import { getAstForSql } from './getAstForSql';

const mysqlSelectStatement = `
  SELECT * FROM \`users\`
  JOIN \`posts\` ON \`users\`.\`id\` = \`posts\`.\`user_id\`
  WHERE \`users\`.\`id\` = 1
  ORDER BY \`users\`.\`id\` ASC
  LIMIT 1
`;

const postgresSelectStatement = `
  SELECT * FROM "users"
  JOIN "posts" ON "users"."id" = "posts"."user_id"
  WHERE "users"."id" = 1
  ORDER BY "users"."id" ASC
  LIMIT 1
`;

describe('getAstForSql', () => {
  describe('parses SQL select statement', () => {
    test('MySQL', () => {
      expect(getAstForSql({ engine: 'mysql', statement: mysqlSelectStatement })).toMatchSnapshot();
    });

    test('PostgreSQL', () => {
      expect(
        getAstForSql({ engine: 'postgresql', statement: postgresSelectStatement }),
      ).toMatchSnapshot();
    });
  });

  test('returns null for invalid SQL', () => {
    expect(getAstForSql({ engine: 'mysql', statement: postgresSelectStatement })).toBeNull();
  });

  test('parses first statement if multiple statements are provided', () => {
    const testKeyword = 'TEST123';

    const ast = getAstForSql({
      engine: 'mysql',
      statement: `${mysqlSelectStatement}; SELECT * FROM \`${testKeyword}\``,
    });

    expect(JSON.stringify(ast)).not.toContain(testKeyword);
  });
});
