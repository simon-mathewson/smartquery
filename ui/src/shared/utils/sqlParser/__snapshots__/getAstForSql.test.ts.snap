// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getAstForSql > parses SQL select statement > MySQL 1`] = `
{
  "columns": [
    {
      "as": null,
      "expr": {
        "column": "*",
        "table": null,
        "type": "column_ref",
      },
    },
  ],
  "distinct": null,
  "from": [
    {
      "as": null,
      "db": null,
      "table": "users",
    },
    {
      "as": null,
      "db": null,
      "join": "INNER JOIN",
      "on": {
        "left": {
          "column": "id",
          "table": "users",
          "type": "column_ref",
        },
        "operator": "=",
        "right": {
          "column": "user_id",
          "table": "posts",
          "type": "column_ref",
        },
        "type": "binary_expr",
      },
      "table": "posts",
    },
  ],
  "groupby": null,
  "having": null,
  "into": {
    "position": null,
  },
  "limit": {
    "seperator": "",
    "value": [
      {
        "type": "number",
        "value": 1,
      },
    ],
  },
  "locking_read": null,
  "options": null,
  "orderby": [
    {
      "expr": {
        "column": "id",
        "table": "users",
        "type": "column_ref",
      },
      "type": "ASC",
    },
  ],
  "type": "select",
  "where": {
    "left": {
      "column": "id",
      "table": "users",
      "type": "column_ref",
    },
    "operator": "=",
    "right": {
      "type": "number",
      "value": 1,
    },
    "type": "binary_expr",
  },
  "window": null,
  "with": null,
}
`;

exports[`getAstForSql > parses SQL select statement > PostgreSQL 1`] = `
{
  "columns": [
    {
      "as": null,
      "expr": {
        "column": "*",
        "table": null,
        "type": "column_ref",
      },
    },
  ],
  "distinct": {
    "type": null,
  },
  "from": [
    {
      "as": null,
      "db": null,
      "table": "users",
    },
    {
      "as": null,
      "db": null,
      "join": "INNER JOIN",
      "on": {
        "left": {
          "column": "id",
          "table": "users",
          "type": "column_ref",
        },
        "operator": "=",
        "right": {
          "column": "user_id",
          "table": "posts",
          "type": "column_ref",
        },
        "type": "binary_expr",
      },
      "table": "posts",
    },
  ],
  "groupby": null,
  "having": null,
  "into": {
    "position": null,
  },
  "limit": {
    "seperator": "",
    "value": [
      {
        "type": "number",
        "value": 1,
      },
    ],
  },
  "options": null,
  "orderby": [
    {
      "expr": {
        "column": "id",
        "table": "users",
        "type": "column_ref",
      },
      "nulls": null,
      "type": "ASC",
    },
  ],
  "type": "select",
  "where": {
    "left": {
      "column": "id",
      "table": "users",
      "type": "column_ref",
    },
    "operator": "=",
    "right": {
      "type": "number",
      "value": 1,
    },
    "type": "binary_expr",
  },
  "window": null,
  "with": null,
}
`;
